### ISSUE TITLE
feat: Implement Tenstorrent Backend Support (TT-NN)

### ISSUE THREAD

**Issue by metaphorics** [cognitive-glitch/whisper.cpp.tt#1]
### Motivation & Context

Tenstorrent offers accessible, high-performance RISC-V based AI accelerators (Grayskull e75/e150 and Wormhole n150/n300). Enabling support for these devices in `whisper.cpp` would provide a high-performance open hardware alternative to existing backends (CUDA, Metal), particularly for edge and server inference.

This proposal outlines the implementation of a new **GGML Backend** (`ggml-ttnn`) leveraging the **TT-Metalium SDK** (specifically the high-level **TT-NN** library).

### Proposed Architecture

The implementation will follow the standard GGML backend interface defined in `ggml-backend.h`.

**Core Design Decisions:**

1.  **Library:** We will target **TT-NN** (Tenstorrent Neural Network), which provides high-level C++ operators (MatMul, Softmax, etc.) optimized for the hardware, avoiding low-level kernel writing.
2.  **Data Layout (The Critical Challenge):**
      * **Problem:** GGML uses **Row-Major** contiguous tensors. Tenstorrent compute units (Tensix cores) operate on **Tiled Layouts** (32x32 blocks) for high performance.
      * **Strategy:** The backend `buffer_type` implementation must transparently handle "Tilization" (Host -> Device) and "Untilization" (Device -> Host).
3.  **Weight Persistence:** Model weights must be uploaded once and persist in Tiled layout on Device DRAM to avoid massive runtime conversion overhead.

-----

#### Phase 1: Build System & Environment

*Objective: Enable compilation with TT-Metalium libraries.*

  - [ ] **Create `cmake/FindTTMetal.cmake`**:
      - Script to locate `$TT_METAL_HOME`, headers (`ttnn/device.hpp`), and libraries (`ttnn`, `tt_metal`, `yaml-cpp`).
  - [ ] **Update Root `CMakeLists.txt`**:
      - Add `WHISPER_TTNN` option (default `OFF`).
  - [ ] **Update `ggml/src/CMakeLists.txt`**:
      - Link against TT libraries if `WHISPER_TTNN` is enabled.
      - **Constraint:** Enforce `CMAKE_CXX_STANDARD 17` when this backend is active (Required by TT-Metalium).
  - [ ] **Docker Support**:
      - Add `.devops/main-ttnn.Dockerfile` based on the official image: `ghcr.io/tenstorrent/tt-metal/tt-metalium-ubuntu-22.04-release-amd64`.

#### Phase 2: Core Backend (`ggml-ttnn.cpp`)

*Objective: Implement the bridge between GGML and Device DRAM.*

  - [ ] **Device Initialization**:
      - Implement `ggml_backend_ttnn_init(device_id)`.
      - **Optimization:** Enable the **Program Cache** to store compiled kernels on disk, preventing long JIT compile times on subsequent runs.
  - [ ] **Buffer Management**:
      - Implement `ggml_backend_ttnn_buffer_type`.
      - **`set_tensor` (Host -> Device)**:
        1.  Create `ttnn::Tensor` from borrowed host pointer (Row-Major).
        2.  Execute `ttnn::to_layout(tensor, ttnn::Layout::TILE)`.
        3.  Execute `ttnn::to_device(tensor, device, DRAM)`.
      - **`get_tensor` (Device -> Host)**:
        1.  Execute `ttnn::from_device(tensor)`.
        2.  Execute `ttnn::to_layout(tensor, ttnn::Layout::ROW_MAJOR)`.
        3.  Copy to host pointer.

#### Phase 3: Operator Mapping

*Objective: Map GGML graph nodes to TT-NN ops.*

| GGML Op | TT-NN Op | Notes |
| :--- | :--- | :--- |
| `GGML_OP_MUL_MAT` | `ttnn::matmul` | **Critical.** Inputs must be Tiled. Supports FP16/BF16. |
| `GGML_OP_ADD` | `ttnn::add` | Element-wise addition. |
| `GGML_OP_GELU` | `ttnn::gelu` | Activation. |
| `GGML_OP_SOFT_MAX` | `ttnn::softmax` | Critical for Attention. |
| `GGML_OP_RMS_NORM` | `ttnn::rms_norm` | Layer Normalization. |
| `GGML_OP_CONV_1D` | `ttnn::conv2d` | Whisper's first layer. Strategy: Reshape input (N, C, L) -> (N, C, 1, L) and use 2D conv, or fallback to CPU. |
| `GGML_OP_GET_ROWS` | `ttnn::embedding` | For token embeddings. |

#### Phase 4: Integration

  - [ ] **Update `src/whisper.cpp`**:
      - Add logic in `whisper_init_state` to initialize the TTNN backend if hardware is present.
  - [ ] **Scheduler Optimization**:
      - Implement `ggml_backend_ttnn_supports_op`.
      - **Crucial:** Return `false` for complex view operations (`RESHAPE`, `PERMUTE`, `TRANSPOSE`) to force CPU fallback. These are expensive to perform on-device without specific kernels, and GGML handles them efficiently on the CPU side via pointer arithmetic.

-----

### Code Scaffolding

**1. CMake Integration (`cmake/FindTTMetal.cmake`)**

```cmake
set(TT_HOME $ENV{TT_METAL_HOME})
find_path(TTNN_INCLUDE_DIR NAMES "ttnn/device.hpp" PATHS ${TT_HOME}/include NO_DEFAULT_PATH)
find_library(TTNN_LIB NAMES "ttnn" PATHS ${TT_HOME}/build/lib NO_DEFAULT_PATH)
find_library(TTMETAL_LIB NAMES "tt_metal" PATHS ${TT_HOME}/build/lib NO_DEFAULT_PATH)

include(FindPackageHandleStandardArgs)
find_package_handle_standard_args(TTMetal DEFAULT_MSG TTNN_LIB TTMETAL_LIB TTNN_INCLUDE_DIR)

if(TTMetal_FOUND)
    set(TT_LIBRARIES ${TTNN_LIB} ${TTMETAL_LIB} yaml-cpp)
    set(TT_INCLUDE_DIRS ${TTNN_INCLUDE_DIR} ${TT_HOME}/tt_metal/include)
endif()

```

**2. Buffer Management Logic (`ggml-ttnn.cpp`)**

```cpp
// Example logic for uploading weights/data
void ttnn_buffer_set_tensor(ggml_backend_buffer_t buffer, ggml_tensor* src, ...) {
    // 1. Wrap host pointer (Row-Major)
    auto host_tensor = ttnn::Tensor::from_borrowed_storage(
        src->data, src->ne, ttnn::Layout::ROW_MAJOR
    );
    
    // 2. Convert to Tiled Layout (Required for Tensix compute)
    auto tiled_tensor = ttnn::to_layout(host_tensor, ttnn::Layout::TILE);
    
    // 3. Move to Device DRAM
    auto device_tensor = ttnn::to_device(tiled_tensor, device_ctx, ttnn::MemoryConfig{DRAM});
    
    // 4. Store 'device_tensor' handle in the backend's internal map
}

```

### Development Environment

The TensTorrent software stack is sensitive to OS and driver versions. Development should occur in the official container:

```bash
docker run -it --rm \
  --device /dev/tenstorrent \
  -v /dev/hugepages-1G:/dev/hugepages-1G \
  -v $(pwd):/app \
  ghcr.io/tenstorrent/tt-metal/tt-metalium-ubuntu-22.04-release-amd64:latest-rc

```

### Tradeoff Considerations

1. **Quantization:** TT-NN relies heavily on BFLOAT16. Initial implementation should focus on F16/BF16. Should we dequantize GGML Q-types to BF16 on the fly during the "Tilization" phase?
2. **KV Cache:** Should the KV cache be allocated in L1 memory (faster) or DRAM (larger capacity)? Given Whisper's size, DRAM is likely safer for initial support.
